jlrpnbbngtn/beer-garden
# Issue/1192 garden connection buttons

### Overview
There are five commits in this PR to be concerned with:

1. **Typo**: A single misspelt word in a Python string. *Safe to ignore.*
2. **Massive eslint refactor**: These changes are *mostly* automatically fixed via `eslint`. Care was taken to ensure the correct version of `eslint` was used with the correct `.eslintrc.json` file. How we came to the place where there were thousands of errors is unknown, but the easy fixes are in place with this commit. There are further corrections that need to be made by hand and will occur in a later commit. *Should be safe to trust*.
3. **Alter UI to work with server data validation**: *recommend testing in conjunction with commit 4*
    a. Testing revealed that `GardenBaseSchema` was returning error messages that were not useful in their current state (strings with single quotes embedded in them don't play well with JSON).
    b. Code was added to disallow updating a garden to have no connection parameters at all.
    c. Code was added to interpret error messages received because of missing non-optional parameters and update the UI with that information.
4. **Add import/export buttons & functionality**: as the name says
5. **Fixed remaining eslint errors**. *Should be safe to trust*.


### Testing
1. We assume testing commences from commit 4 or 5. Using the usual methods, start parent and child instances of beer garden.
2. Access the garden view page for the child garden.
3. Set the connection type to *HTTP* and update the *Host Name* and  *Port* fields to their correct values (e.g., *bg-child1* and *2447*).
4. Select the *Export Connection* button. Examine the JSON file saved to the local file system to verify it is a faithful representation of the updates to the form.
5. Delete all fields for HTTP (*Host Name*, *Port* and *URL Prefix*). Press the *Save Configuration* button. Note that the UI updates with an appropriate error message about not allowing both HTTP and Stomp configuration to be empty.
6. Select the *Import Connection* button and chose the JSON file that was just saved. Note that the UI updates with the previous values for the connection.
7. Again select the "Save Configuration" button. Note that the previous error message is cleared. Use the "Sync" button to verify that the garden configuration has been updated on the server (or examine the database directly).
8. Set the *Port* to a disallowed value (e.g. zero or a negative number). Again select the "Save Configuration" button. Note that the UI updates to show that there is an error in the HTTP configuration (this would be useful if you had selected the STOMP tab before pressing the button). Note that on the HTTP tab there is a validation error associated with the port field and that it communicates the error actually received from the server (*Value out of range for ports* in this case).
9. Delete the *Port* field completely and choose the "Save Configuration" button again. Note that the invalidation message associated with the port field updates.
10. Reset the *Port* field to a legal value (e.g., *2447*). Set a single optional value on the Stomp tab. For example, set the *Send Destination* to a non-empty string. Choose the "Save Configuration" button again. Note that UI updates to alert of the stomp connection errors and the problematic fields are highlighted on the STOMP tab.
11. Delete the previously set stomp field so that every field on the stomp tab is blank. Select the *Connection Type* as *STOMP* and press the "Save Configuration" button again. Verify that the UI updates with a message indicating an empty connection is not allowed.
12. *Optional*: Alter the saved JSON file and reload it using the button. Verify that only the UI <ins>form</ins> is updated. In other words, there is no attempt to update the server until the "Save Configuration" button is pressed. This behavior is different from the *Import Jobs* functionality.


### Other
1. Leaving the host name blank on the http tab doesn't always produce an error. In some cases it causes the server to throw an exception (`AttributeError: 'dict' object has no attribute 'id'` from a malformed call to a logger) and in other cases it updates without issue *with the disallowed blank field*. Other times everything works as expected. This behavior is believed to not be related to this work and will be investigated under a different ticket.
